@switch exhaustiveness does not work for @for loop nad @if variables

### Which @angular/* package(s) are the source of the bug?

compiler

### Is this a regression?

No

### Description

Summary
`@switch` exhaustiveness checks do not work when switching on a variable introduced by `@for` or `@if`.

Example

```html
@for (nested of nesteds; track nested) {
  @switch (nested.type) { 
    @case ('alpha') { 
      <span>{{ nested.alpha }}</span> 
    } 
    @case ('beta') { 
      <span>{{ nested.beta }}</span> 
    } 
    @default never; 
  }
}

@if (optional.maybe; as union) {
  @switch (union.type) { 
    @case ('alpha') { 
      <span>{{ union.alpha }}</span> 
    } 
    @case ('beta') { 
      <span>{{ union.beta }}</span> 
    }
    @default never; 
  }
}
```

Expected behavior
`@switch` exhaustiveness should work for `@for` loop / `@if` variables (including their discriminant fields), the same way it works for direct const references, when the switched value is a finite/discriminated union.

Actual behavior
The compilator reports an error.

### Please provide a link to a minimal reproduction of the bug

https://stackblitz.com/edit/stackblitz-starters-4ticm6ar

### Please provide the exception or error you saw

```true
✘ [ERROR] TS2322: Type 'any' is not assignable to type 'never'. [plugin angular-compiler]

✘ [ERROR] TS2339: Property 'kind' does not exist on type 'never'. [plugin angular-compiler]
```

### Please provide the environment you discovered this bug in (run `ng version`)

```true
_                      _                 ____ _     ___
    / \   _ __   __ _ _   _| | __ _ _ __     / ___| |   |_ _|
   / △ \ | '_ \ / _` | | | | |/ _` | '__|   | |   | |    | |
  / ___ \| | | | (_| | |_| | | (_| | |      | |___| |___ | |
 /_/   \_\_| |_|\__, |\__,_|_|\__,_|_|       \____|_____|___|
                |___/
    

Angular CLI       : 21.2.0
Angular           : 21.2.0
Node.js           : 24.13.0
Package Manager   : bun 1.3.10
Operating System  : darwin arm64

┌──────────────────────────────────┬───────────────────┬───────────────────┐
│ Package                          │ Installed Version │ Requested Version │
├──────────────────────────────────┼───────────────────┼───────────────────┤
│ @angular-devkit/build-angular    │ 21.2.0            │ 21.2.0            │
│ @angular-devkit/core             │ 21.2.0            │ 21.2.0            │
│ @angular-devkit/schematics       │ 21.2.0            │ 21.2.0            │
│ @angular/animations              │ 21.2.0            │ 21.2.0            │
│ @angular/cdk                     │ 21.2.0            │ 21.2.0            │
│ @angular/cli                     │ 21.2.0            │ 21.2.0            │
│ @angular/common                  │ 21.2.0            │ 21.2.0            │
│ @angular/compiler                │ 21.2.0            │ 21.2.0            │
│ @angular/compiler-cli            │ 21.2.0            │ 21.2.0            │
│ @angular/core                    │ 21.2.0            │ 21.2.0            │
│ @angular/elements                │ 21.2.0            │ 21.2.0            │
│ @angular/forms                   │ 21.2.0            │ 21.2.0            │
│ @angular/language-service        │ 21.2.0            │ 21.2.0            │
│ @angular/localize                │ 21.2.0            │ 21.2.0            │
│ @angular/material                │ 21.2.0            │ 21.2.0            │
│ @angular/material-luxon-adapter  │ 21.2.0            │ 21.2.0            │
│ @angular/material-moment-adapter │ 21.2.0            │ 21.2.0            │
│ @angular/platform-browser        │ 21.2.0            │ 21.2.0            │
│ @angular/platform-server         │ 21.2.0            │ 21.2.0            │
│ @angular/router                  │ 21.2.0            │ 21.2.0            │
│ @angular/service-worker          │ 21.2.0            │ 21.2.0            │
│ @schematics/angular              │ 21.2.0            │ 21.2.0            │
│ ng-packagr                       │ 21.2.0            │ 21.2.0            │
│ rxjs                             │ 7.8.2             │ 7.8.2             │
│ typescript                       │ 5.9.3             │ 5.9.3             │
│ vitest                           │ 4.0.18            │ 4.0.18            │
└──────────────────────────────────┴───────────────────┴───────────────────┘
```

### Anything else?

_No response_

--- a/adev/src/content/guide/templates/control-flow.md
+++ b/adev/src/content/guide/templates/control-flow.md
@@ -160,3 +160,20 @@ export class AppComponent {
   state: 'loggedOut' | 'loading' | 'loggedIn' = 'loggedOut';
 }
 ```
+
+When the switched expression is nested within a union, you must explicitly specify the expression to check for exhaustiveness.
+
+```angular-ts
+@Component({
+  template: `
+    @switch (state.mode) {
+      @case ('show') { {{ state.menu }}; }
+      @case ('hide') {}
+      @default never(state);
+    }
+  `,
+})
+export class App {
+  state!: {mode: 'hide'} | {mode: 'show'; menu: number};
+}
+```

--- a/packages/compiler-cli/src/ngtsc/typecheck/src/ops/switch_block.ts
+++ b/packages/compiler-cli/src/ngtsc/typecheck/src/ops/switch_block.ts
@@ -60,7 +60,12 @@ export class TcbSwitchOp extends TcbOp {
     });
 
     if (this.block.exhaustiveCheck) {
-      const switchValue = tcbExpression(this.block.expression, this.tcb, this.scope);
+      let translateExpression = this.block.expression;
+      if (this.block.exhaustiveCheck.expression) {
+        translateExpression = this.block.exhaustiveCheck.expression;
+      }
+
+      const switchValue = tcbExpression(translateExpression, this.tcb, this.scope);
       const exhaustiveId = this.tcb.allocateId();
 
       clauses.push(

--- a/packages/compiler-cli/src/ngtsc/typecheck/test/diagnostics_spec.ts
+++ b/packages/compiler-cli/src/ngtsc/typecheck/test/diagnostics_spec.ts
@@ -1453,6 +1453,25 @@ class TestComponent {
 
       expect(messages).toEqual([]);
     });
+
+    it('should narrow union when switching on a nested prop', () => {
+      const messages = diagnose(
+        `
+        @switch (state.mode) {
+          @case ('show') { {{ state.menu }}; }
+          @case ('hide') {}
+          @default never(state);
+        }
+        `,
+        `
+          export class TestComponent {
+            state: { mode: 'hide' } | { mode: 'show'; menu: number };
+          }
+        `,
+      );
+
+      expect(messages).toEqual([]);
+    });
   });
 
   // https://github.com/angular/angular/issues/43970

--- a/packages/compiler-cli/src/ngtsc/typecheck/test/type_check_block_spec.ts
+++ b/packages/compiler-cli/src/ngtsc/typecheck/test/type_check_block_spec.ts
@@ -2202,6 +2202,27 @@ describe('type check blocks', () => {
       );
     });
 
+    it('should generate a switch block with exhaustiveness checking with param', () => {
+      const TEMPLATE = `
+        @switch (expr) {
+          @case (1) {
+            {{one()}}
+          }
+          @case (2) {
+            {{two()}}
+          }
+          @default never(expr.prop);
+        }
+      `;
+
+      expect(tcb(TEMPLATE)).toContain(
+        'switch (((this).expr)) { ' +
+          'case 1: "" + ((this).one()); break; ' +
+          'case 2: "" + ((this).two()); break; ' +
+          'default: const tcbExhaustive_t1: never = ((((this).expr)).prop);',
+      );
+    });
+
     it('should not report unused locals for exhaustiveness check variable', () => {
       const TEMPLATE = `
         @switch (expr) {

--- a/packages/compiler/src/ml_parser/lexer.ts
+++ b/packages/compiler/src/ml_parser/lexer.ts
@@ -299,14 +299,6 @@ class _Tokenizer {
     this._beginToken(TokenType.BLOCK_OPEN_START, start);
     const startToken = this._endToken([this._getBlockName()]);
 
-    if (startToken.parts[0] === 'default never' && this._attemptCharCode(chars.$SEMICOLON)) {
-      this._beginToken(TokenType.BLOCK_OPEN_END);
-      this._endToken([]);
-      this._beginToken(TokenType.BLOCK_CLOSE);
-      this._endToken([]);
-      return;
-    }
-
     if (this._cursor.peek() === chars.$LPAREN) {
       // Advance past the opening paren.
       this._cursor.advance();
@@ -322,6 +314,14 @@ class _Tokenizer {
         startToken.type = TokenType.INCOMPLETE_BLOCK_OPEN;
         return;
       }
+    }
+
+    if (startToken.parts[0] === 'default never' && this._attemptCharCode(chars.$SEMICOLON)) {
+      this._beginToken(TokenType.BLOCK_OPEN_END);
+      this._endToken([]);
+      this._beginToken(TokenType.BLOCK_CLOSE);
+      this._endToken([]);
+      return;
     }
 
     if (this._attemptCharCode(chars.$LBRACE)) {

--- a/packages/compiler/src/render3/r3_ast.ts
+++ b/packages/compiler/src/render3/r3_ast.ts
@@ -471,6 +471,7 @@ export class SwitchBlockCaseGroup extends BlockNode implements Node {
 
 export class SwitchExhaustiveCheck extends BlockNode implements Node {
   constructor(
+    public expression: AST | null,
     sourceSpan: ParseSourceSpan,
     startSourceSpan: ParseSourceSpan,
     endSourceSpan: ParseSourceSpan | null,

--- a/packages/compiler/src/render3/r3_control_flow.ts
+++ b/packages/compiler/src/render3/r3_control_flow.ts
@@ -264,6 +264,10 @@ export function createSwitchBlock(
     if (isCase) {
       expression = parseBlockParameterToBinding(node.parameters[0], bindingParser);
     } else if (node.name === 'default never') {
+      if (node.parameters.length > 0) {
+        expression = parseBlockParameterToBinding(node.parameters[0], bindingParser);
+      }
+
       if (
         node.children.length > 0 ||
         (node.endSourceSpan !== null &&
@@ -287,6 +291,7 @@ export function createSwitchBlock(
       }
 
       exhaustiveCheck = new t.SwitchExhaustiveCheck(
+        expression,
         node.sourceSpan,
         node.startSourceSpan,
         node.endSourceSpan,

--- a/packages/compiler/src/render3/view/t2_binder.ts
+++ b/packages/compiler/src/render3/view/t2_binder.ts
@@ -953,7 +953,7 @@ class TemplateBinder extends CombinedRecursiveAstVisitor {
   }
 
   override visitSwitchExhaustiveCheck(block: SwitchExhaustiveCheck) {
-    // There are no bindings/references in the exhaustive check block.
+    block.expression?.visit(this);
   }
 
   override visitForLoopBlock(block: ForLoopBlock) {

--- a/packages/compiler/test/ml_parser/lexer_spec.ts
+++ b/packages/compiler/test/ml_parser/lexer_spec.ts
@@ -3396,6 +3396,16 @@ describe('HtmlLexer', () => {
       ]);
     });
 
+    it('should parse @default never(expr);', () => {
+      expect(tokenizeAndHumanizeParts('@default never(expr);')).toEqual([
+        [TokenType.BLOCK_OPEN_START, 'default never'],
+        [TokenType.BLOCK_PARAMETER, 'expr'],
+        [TokenType.BLOCK_OPEN_END],
+        [TokenType.BLOCK_CLOSE],
+        [TokenType.EOF],
+      ]);
+    });
+
     it('should parse @default never ;', () => {
       expect(tokenizeAndHumanizeParts('@default never ;')).toEqual([
         [TokenType.BLOCK_OPEN_START, 'default never'],

--- a/packages/language-service/src/template_target.ts
+++ b/packages/language-service/src/template_target.ts
@@ -677,7 +677,9 @@ class TemplateTargetVisitor implements TmplAstVisitor {
     this.visitAll(block.children);
   }
 
-  visitSwitchExhaustiveCheck(block: TmplAstSwitchExhaustiveCheck) {}
+  visitSwitchExhaustiveCheck(block: TmplAstSwitchExhaustiveCheck) {
+    block.expression && this.visitBinding(block.expression);
+  }
 
   visitForLoopBlock(block: TmplAstForLoopBlock) {
     this.visit(block.item);

--- a/tools/manual_api_docs/blocks/switch.md
+++ b/tools/manual_api_docs/blocks/switch.md
@@ -57,3 +57,20 @@ export class AppComponent {
   state: 'loggedOut' | 'loading' | 'loggedIn' = 'loggedOut';
 }
 ```
+
+When the switched expression is nested within a union, you must explicitly specify the expression to check for exhaustiveness.
+
+```angular-ts
+@Component({
+  template: `
+    @switch (state.mode) {
+      @case ('show') { {{ state.menu }}; }
+      @case ('hide') {}
+      @default never(state);
+    }
+  `,
+})
+export class App {
+  state!: {mode: 'hide'} | {mode: 'show'; menu: number};
+}
+```

--- a/vscode-ng-language-service/syntaxes/test/data/template-blocks.html
+++ b/vscode-ng-language-service/syntaxes/test/data/template-blocks.html
@@ -23,6 +23,11 @@
     @default never;
 }
 
+@switch(aOrb) {
+    @case(a) {}
+    @default never(aOrb.c);
+}
+
 @if (a==b) { hello } @else  { goodbye }
 
 @if (a==b) { 

